db/batchlog_manager.cc

```diff
@@ -189,20 +189,14 @@ future<> db::batchlog_manager::replay_all_failed_batches() {
         auto size = data.size();
 
         return map_reduce(*fms, [this, written_at] (canonical_mutation& fm) {
-            return _sys_ks.get_truncated_at(fm.column_family_id()).then([written_at, &fm] (db_clock::time_point t) ->
-                    std::optional<std::reference_wrapper<canonical_mutation>> {
-                if (written_at > t) {
-                    return { std::ref(fm) };
-                } else {
-                    return {};
-                }
-            });
+            const auto& cf = _qp.proxy().local_db().find_column_family(fm.column_family_id());
+            return make_ready_future<canonical_mutation*>(written_at > cf.get_truncation_time() ? &fm : nullptr);
```

Comment:
This change deserves a more descriptive patch commit message saying why exactly truncation record can be safely and reliably obtained from cf itself rather than from sysks.

Presumably the "why" is because truncation records are loaded from sysks and are cached on cfs _before_ this loop executes. If that's the case there should be something that programmatically checks that runtime. Otherwise one day this sequence would change and we won't notice it until some obscure failure in some longevity test.

Reply:
Agree. We should probably make `table::_truncated_at` optional and generate an error if it's accessed before initialisation. I'll update the PR.